oss_src/fileio/libhdfs_shim.cpp

```diff
@@ -384,6 +385,26 @@ extern  "C" {
     search_prefixes = {""};
     search_suffixes = {""};
     file_name = "libjvm.dylib";
+
+    // Run /usr/libexec/java_home to get the libjvm location
+    std::string libjvm_location = "";
+    std::string java_home_cmd = "/usr/libexec/java_home";
+    try {
+      graphlab::process p;
+      p.popen(java_home_cmd,
+              std::vector<std::string>(),
+              STDOUT_FILENO);
+      libjvm_location = p.read_from_child();
+      logstream(LOG_INFO) << "Obtain JAVA_HOME from " << java_home_cmd << ": " << libjvm_location << std::endl;
+    } catch (...) {
+      logstream(LOG_WARNING) << "Error running " << java_home_cmd << std::endl;
+      libjvm_location = "";
+    }
+    if (!libjvm_location.empty()) {
+      // Make this location to be searched first `/usr/libexec/java_home`/jre/lib/server/libjvm.dylib
+      search_prefixes.insert(search_prefixes.begin(), libjvm_location);
```

Comment:
I believe this does the right thing.

The output of /usr/libexec/java_home is used as `search_prefix`.
'/jre/lib/server' is used as `search_suffix`.
'libjvm.dylib' is the `filen_ame`.

At the end of the function, the potential search paths are generated by
doing combinations of search_prefix / search_suffix / file_name

__

<http://www.dato.com/>   Jay (Haijie) Gu
CoFounder & Principal Data Scientist
jgu@dato.com <name@dato.com>   <https://twitter.com/datoinc>
<http://www.linkedin.com/company/graphlab>
<https://www.facebook.com/datoinc>
<http://www.youtube.com/channel/UC9T07WMdwB6IPf_dKXhVsCQ>


On Wed, Feb 3, 2016 at 2:27 PM, Yucheng Low <notifications@github.com>
wrote:

> In oss_src/fileio/libhdfs_shim.cpp
> <https://github.com/dato-code/SFrame/pull/160#discussion_r51798227>:
>
> > +    std::string libjvm_location = "";
> > +    std::string java_home_cmd = "/usr/libexec/java_home";
> > +    try {
> > +      graphlab::process p;
> > +      p.popen(java_home_cmd,
> > +              std::vector<std::string>(),
> > +              STDOUT_FILENO);
> > +      libjvm_location = p.read_from_child();
> > +      logstream(LOG_INFO) << "Obtain JAVA_HOME from " << java_home_cmd << ": " << libjvm_location << std::endl;
> > +    } catch (...) {
> > +      logstream(LOG_WARNING) << "Error running " << java_home_cmd << std::endl;
> > +      libjvm_location = "";
> > +    }
> > +    if (!libjvm_location.empty()) {
> > +      // Make this location to be searched first `/usr/libexec/java_home`/jre/lib/server/libjvm.dylib
> > +      search_prefixes.insert(search_prefixes.begin(), libjvm_location);
>
> ah. the bug report was ambiguous. it is not " /usr/libexec/java_home" or
> "jre/lib/server/libjvm.dylib".
> where x = /usr/libexec/java_home
> the path is "$x/jre/lib/server/libjvm.dylib"
>
> See the shell script in the bug report
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dato-code/SFrame/pull/160/files#r51798227>.
>